#VRML_SIM R2025a utf8

EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "https://raw.githubusercontent.com/cyberbotics/webots/R2025a/projects/objects/floors/protos/RectangleArena.proto"

WorldInfo {
  contactProperties [
    ContactProperties {
      material1 "rubber"
      material2 "ground"
      coulombFriction [
        0.8
      ]
    }
    ContactProperties {
      material1 "rubber"
      material2 "whiteboard"
      coulombFriction [
        0.7
      ]
    }
    ContactProperties {
      material1 "rubber"
      material2 "bumpGround"
      coulombFriction [
        0.85
      ]
    }
  ]
}
Viewpoint {
  orientation -0.22621903483761696 0.3997686399970203 0.8882623389247513 1.054045364897764
  position -2.0667200369226086 -2.96126071140145 1.915006017319906
}
TexturedBackground {
}
TexturedBackgroundLight {
}
Viewpoint {
  orientation -0.2256052783179886 0.733251589159842 0.6414392920531368 1.148561532759604
  position -2.480124444647765 -1.7721924346661442 3.1127122535974543
}
TexturedBackground {
}
TexturedBackgroundLight {
  luminosity 0
}
RectangleArena {
  contactMaterial "cardboard"
  floorSize 3 3
}
DEF WAYPOINT_MARKER Solid {
  translation 0.3832 -0.0597 0.002
  children [
    Shape {
      appearance PBRAppearance {
        baseColor 1 0 0
        transparency 0.8
        roughness 1
        metalness 0
        emissiveColor 1 0 0
      }
      geometry Cylinder {
        height 1
        radius 0.03
      }
    }
  ]
}
DEF POLULU_ROBOT Robot {
  translation 0.2733453195725872 -0.07896910302096248 0.016358063554386692
  rotation -0.026503328509118136 -0.02784429976113839 0.9992608611112265 1.6612432461789877
  children [
    Display {
      width 230
    }
    InertialUnit {
      name "imu"
    }
    GPS {
    }
    Shape {
      appearance PBRAppearance {
        roughness 1
        metalness 0
      }
      geometry Mesh {
        url [
          "../../../Downloads/bodywithoutcaster.obj"
        ]
      }
      castShadows FALSE
      isPickable FALSE
    }
    BallJoint {
      jointParameters BallJointParameters {
        position -1366.0415451093936
        anchor 0 0.0404 -0.0096
      }
      jointParameters2 JointParameters {
        position 24.0675277061411
        axis 1 0 0
      }
      jointParameters3 JointParameters {
        position 21.661988241580094
        axis 0 1 0
      }
      endPoint Solid {
        translation 0.02784092099824284 0.04874431163069451 -0.03953713861881482
        rotation 0.8145922943079033 -0.568833463017276 -0.1134367021997706 2.5174814880784524
        children [
          Shape {
            appearance PBRAppearance {
              roughness 1
              metalness 0
            }
            geometry Mesh {
              url [
                "../Models/caster.obj"
              ]
            }
            castShadows FALSE
            isPickable FALSE
          }
        ]
        name "caster"
        boundingObject Pose {
          translation 0 0.0404 -0.00963
          children [
            Sphere {
              radius 0.00635
              subdivision 3
            }
          ]
        }
        physics Physics {
        }
        linearVelocity 0.19914650237029674 -0.0049477683955082424 0.023775077955968428
        angularVelocity 0.7791727213010251 31.361645374422473 -3.1118520081274434
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        position 2251.3502530968535
      }
      device [
        RotationalMotor {
          name "LeftMotor"
          maxVelocity 75.4
        }
      ]
      endPoint Solid {
        translation 0.041499474598696594 9.627159584648691e-05 -0.0003584238673708846
        rotation -0.6697355373276016 0.32451731321299837 0.6679392363572642 3.77857168500624
        children [
          DEF wheel Shape {
            appearance PBRAppearance {
              roughness 1
              metalness 0
            }
            geometry Cylinder {
              height 0.0068
              radius 0.016
              subdivision 500
            }
            castShadows FALSE
            isPickable FALSE
          }
        ]
        contactMaterial "rubber"
        boundingObject USE wheel
        physics Physics {
        }
        linearVelocity 0.13955046389167186 0.009695936694695265 -1.921160328422231e-05
        angularVelocity -0.6170255802468303 8.590703441462122 0.5864476205723991
      }
    }
    HingeJoint {
      jointParameters HingeJointParameters {
        position 3126.543930395829
      }
      device [
        DEF RightMotor RotationalMotor {
          name "RightMotor"
          maxVelocity 75.4
        }
      ]
      endPoint Solid {
        translation -0.04149868078943604 5.175274968812217e-05 -0.0005212320644572875
        rotation 0.6890061809831222 -0.23405368909050736 0.6859222646846924 3.605533370468954
        children [
          USE wheel
        ]
        contactMaterial "rubber"
        boundingObject USE wheel
        physics Physics {
        }
        linearVelocity 0.18776957478190687 0.013601527253330206 -5.204255236288087e-06
        angularVelocity -0.8387584748212309 11.598553275608026 0.5977122828272619
      }
    }
  ]
  boundingObject Pose {
    translation 0 0 0.0025
    children [
      Cylinder {
        height 0.0266
        radius 0.0484
        subdivision 50
      }
    ]
  }
  physics Physics {
    mass 0.158
    centerOfMass [
      0 0 0
    ]
  }
  controller "autonomous_navigation"
  supervisor TRUE
  linearVelocity 0.1685642811661311 0.011965580902264926 0.000899043319076709
  angularVelocity 0.060570139715621515 -1.187861464609532 0.5893038834206932
}
Solid {
  translation -0.34 1.01 5.00225e-06
  rotation 0.5773509358554485 0.5773489358556708 0.5773509358554485 2.0944
  children [
    DEF obstacle Shape {
      appearance PBRAppearance {
        baseColor 0 0 0
        roughness 1
        metalness 0
      }
      geometry Mesh {
        url [
          "../Models/ramp.obj"
        ]
      }
    }
  ]
  name "ramp"
  contactMaterial "whiteboard"
  boundingObject Group {
    children [
      Pose {
        translation -0.590961 0.033 0
        rotation 0 0 1 0.3321694619948667
        children [
          Box {
            size 0.76 0.1 0.28
          }
        ]
      }
      Pose {
        translation -0.01 0.1024 0
        children [
          Box {
            size 0.25 0.25 0.31
          }
        ]
      }
      Pose {
        translation -0.108 -0.2061 0
        children [
          Capsule {
            height 0.4
            radius 0.434
            subdivision 500
          }
        ]
      }
    ]
  }
}
Solid {
  translation 0.58 0.99 3.36042e-06
  rotation 0.5773509358554485 0.5773489358556708 0.5773509358554485 2.0944
  children [
    DEF bump Shape {
      appearance PBRAppearance {
        baseColor 0 0 0
        roughness 1
        metalness 0
      }
      geometry Mesh {
        url [
          "../../../Downloads/bump.obj"
        ]
      }
    }
  ]
  name "bump"
  contactMaterial "bumpGround"
  boundingObject Group {
    children [
      Pose {
        translation -0.97 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.97 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.97 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.97 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.97 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.97 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.91 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.91 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.91 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.91 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.91 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.91 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.85 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.85 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.85 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.85 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.85 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.85 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.79 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.79 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.79 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.79 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.79 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.79 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.73 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.73 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.73 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.73 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.73 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.73 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.67 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.67 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.67 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.67 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.67 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.67 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.61 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.61 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.61 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.61 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.61 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.61 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.55 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.55 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.55 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.55 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.55 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.55 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.49 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.49 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.49 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.49 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.49 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.49 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.43 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.43 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.43 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.43 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.43 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.43 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.37 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.37 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.37 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.37 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.37 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.37 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.31 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.31 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.31 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.31 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.31 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.31 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.25 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.25 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.25 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.25 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.25 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.25 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.19 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.19 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.19 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.19 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.19 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.19 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.13 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.13 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.13 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.13 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.13 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.13 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.07 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.07 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.07 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.07 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.07 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.07 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.01 0.0052 -0.01
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.01 0.0052 -0.07
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.01 0.0052 -0.13
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.01 0.0052 -0.19
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.01 0.0052 -0.25
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.01 0.0052 -0.31
        children [
          Box {
            size 0.02 0.004 0.02
          }
        ]
      }
      Pose {
        translation -0.500961 0.0033 -0.17
        rotation 0 0 -1 0
        children [
          Box {
            size 1 0.0034 0.34
          }
        ]
      }
    ]
  }
}
